pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -177,267 +186,602 @@ public synchronized void addDataType(String type, Class<? extends PGobject> klas
+  leverage the fact that typlen = -1 (variable length) for arrays, but that will run afoul with
+  element types that have multiple array and vector implementations such as int2, which has _int2 and
+  int2vector. The _int2 type uses array_in and corresponds to int2[] and smallint[], while the
+  int2vector type uses int2vectorin and is used in system tables.
+
+  The pg_type.typarray column, added in 8.3, makes this unnecessary, so the queries can be updated
+  when 8.2 support is dropped. The efficiencies gained in query performance are negligible in
+  comparison to time spent over the wire and the overhead in maintaining a separate code path for 8.2.
+   */
 
-    if (dotIndex == -1 && !hasQuote && !isArray) {
+  private PreparedStatement getOidStatement(ParsedTypeName typeName) throws SQLException {
+    if (typeName.isSimple()) {
       if (_getOidStatementSimple == null) {
-        String sql;
-        // see comments in @getSQLType()
-        // -- go with older way of unnesting array to be compatible with 8.0
-        sql = "SELECT pg_type.oid, typname "
-              + "  FROM pg_catalog.pg_type "
-              + "  LEFT "
-              + "  JOIN (select ns.oid as nspoid, ns.nspname, r.r "
-              + "          from pg_namespace as ns "
-              + "          join ( select s.r, (current_schemas(false))[s.r] as nspname "
-              + "                   from generate_series(1, array_upper(current_schemas(false), 1)) as s(r) ) as r "
-              + "         using ( nspname ) "
-              + "       ) as sp "
-              + "    ON sp.nspoid = typnamespace "
-              + " WHERE typname = ? "
-              + " ORDER BY sp.r, pg_type.oid DESC LIMIT 1;";
+        /*
+        For bug-compatibility reasons, this query does not restrict the search to the search path.
+        See https://github.com/pgjdbc/pgjdbc/commit/b383f6d2c8f19e2b5b867039ca96071ba8d495e1
+        */
+        String sql = "SELECT t.oid, n.nspname = ANY(current_schemas(true)), n.nspname,"
+            + " t.typinput = 'array_in'::regproc,"
+            + " t.typname, t.typelem, t.typtype, t.typdelim,"
+            + " e.typname, e.typtype, e.typdelim,"
+            + " arr.oid, e.typtype"
+            + "  FROM pg_catalog.pg_type t"
+            + "  JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace"
+            + "  LEFT JOIN pg_catalog.pg_type arr ON (arr.typelem, arr.typinput) = (t.oid, 'array_in'::regproc)"
+            + "  LEFT JOIN pg_catalog.pg_type e ON t.typelem = e.oid"
+            + "  LEFT JOIN (SELECT s.r, (current_schemas(false))[r] AS nspname"
+            + "               FROM generate_series(1, array_upper(current_schemas(false), 1)) AS s (r)) AS sp"
+            + "    USING (nspname)"
+            + " WHERE t.typname = ?"
+            + " ORDER BY sp.r, t.oid DESC LIMIT 1;";
         _getOidStatementSimple = _conn.prepareStatement(sql);
```

Comment:
does this assignment need to be synchronized?

Reply:
As you rightly pointed out below, these private methods are only called by public method which is already synchronized. I've added comments to this effect.